presto-iceberg/src/main/java/io/prestosql/plugin/iceberg/IcebergMetadata.java

```diff
@@ -153,10 +157,18 @@ public IcebergTableHandle getTableHandle(ConnectorSession session, SchemaTableNa
     public Optional<SystemTable> getSystemTable(ConnectorSession session, SchemaTableName tableName)
     {
         IcebergTableHandle table = IcebergTableHandle.from(tableName);
-        if (table.getTableType() == TableType.PARTITIONS) {
-            org.apache.iceberg.Table icebergTable = getIcebergTable(metastore, hdfsEnvironment, session, table.getSchemaTableName());
-            SystemTable systemTable = new PartitionTable(table, session, typeManager, icebergTable);
-            return Optional.of(new ClassLoaderSafeSystemTable(systemTable, getClass().getClassLoader()));
+        return getRawSystemTable(session, table)
```

Comment:
```suggestion
        IcebergTableHandle tableHandle = IcebergTableHandle.from(tableName);
        org.apache.iceberg.Table icebergTable = getIcebergTable(metastore, hdfsEnvironment, session, tableHandle.getSchemaTableName());
        SystemTable systemTable = null;
        switch (tableHandle.getTableType()) {
            case PARTITIONS:
                systemTable = new PartitionTable(tableHandle, session, typeManager, icebergTable);
                break;
            case HISTORY:
                systemTable = new HistoryTable(tableHandle.getSchemaTableName(), icebergTable);
                break;
        }
        return Optional.ofNullable(systemTable).map(ClassLoaderSafeSystemTable::new);
```
and introduce constructor for `ClassLoaderSafeSystemTable` that uses `delegate.getClass().getClassLoader()`.

Reply:
I'd rather not introduce such a constructor for `ClassLoaderSafeSystemTable`. Otherwise it won't be consistent with other `ClassLoaderSafexxx` wrappers. Also I think it doesn't hurt to use a helper function so I'd rather keep it.

Reply:
While helper function does not hurt it also does not help and its introduction is not well justified, IMO. If you are against another constructor for `ClassLoaderSafeSystemTable`, function reference can't be used, but it will be more accurate still to use `systemTable` class loader (compared to `IcebergMetadata` class loader) as all calls are delegated to it.

Reply:
I think explicitly passing in the class loader makes the code clearer to the reader.

We know the class loader model for plugins is a single class loader for the entire plugin, so this is safe. The reason for setting the class loader is actually for Hadoop code, so `IcebergMetadata` and `systemTable` could both be considered right and wrong, depending on your viewpoint. You could argue that `org.apache.hadoop.fs.FileSystem` might be a better choice, but all that matters is we use some class from the plugin to get the plugin's class loader.

As for the helper method, that was my suggestion. Using a method makes the switch code more concise, since you can return directly rather than needing to break (like the upcoming switch expressions language feature), which improves readability.

Reply:
IMO, as the code relies on a single class loader model, providing a reasonable default in `ClassLoaderSafeSystemTable` constructor for the `classLoader` argument makes it (model) more explicit and correspondently code more readable and also easier to write. My guess is that such default would work in 80% (if not more) cases when `ClassLoaderSafeSystemTable` is constructed, so the requirement to provide `classLoader` explicitly is excessive.

I agree that the upcoming switch expression will provide the most readable option, but until it is introduced both options have cons. Should you decide to keep the helper function, I would suggest to rename it and re-write as
```suggestion
    private SystemTable newSystemTableFromType(ConnectorSession session, IcebergTableHandle table)
    {
        org.apache.iceberg.Table icebergTable = getIcebergTable(metastore, hdfsEnvironment, session, table.getSchemaTableName());
        switch (table.getTableType()) {
            case PARTITIONS:
                return new PartitionTable(table, session, typeManager, icebergTable);
            case HISTORY:
                return new HistoryTable(table.getSchemaTableName(), icebergTable);
            default:
                return null;
        }
    }        
```

Reply:
Actually, the current naming and return type look fine to me. What do you think? @electrum

Reply:
To clarify, changing the return type is not a goal by itself, it helps avoiding calls to `Optional` in multiple `switch` branches and to return directly from `case`. I would also suggest making the helper function a method of `TableType`.